src/gtk/toga_gtk/widgets/canvas.py

```diff
@@ -33,69 +28,51 @@ def create(self):
 
         self.native = Gtk.DrawingArea()
         self.native.interface = self.interface
-        surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, self.native.get_allocated_width(),
-                                     self.native.get_allocated_height())
-        self.native.context = cairo.Context(surface)
+        self.surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, self.native.get_allocated_width(),
+                                          self.native.get_allocated_height())
+        self.native_context = cairo.Context(self.surface)
         self.native.font = None
 
-    def set_on_draw(self, handler):
-        self.native.connect('draw', handler)
-
-    def set_context(self, context):
-        self.native.context = context
+    def context(self, context):
+        def draw_callback(canvas, native_context):
+            self.native_context = native_context
+            for drawing_object in traverse(context.drawing_objects):
+                drawing_object(self)
+        self.native.connect('draw', draw_callback)
```

Comment:
Does this mean you're creating a separate draw callback handler for every context? Can't we use a single callback that just draws itself (and it's children)?